### ISSUE TITLE
Float converting to Integer in Expression language

### ISSUE THREAD

**Issue by achandana** [telefonicaid/iotagent-node-lib#868]
I am using the jexl expression language in my device provisioning as below :
 

>  {
>          "devices": [
>          {
>           "device_id": "test1",
>           "entity_name": "urn:ngsi-ld:Motion:test1",
>           "entity_type": "Motion",
>           "attributes": [
>             {
>               "object_id": "Air_Pressure",
>               "name": "Air_Pressure",
>               "type": "Number",
>               "expression":"Air_Pressure + 'unitCode:CAL,SetPoint_Min:-1,SetPoint_Max:1'"
>             },
>             {
>               "object_id": "Humidity",
>               "name": "Humidity",
>               "type": "Number",
>               "expression":"Humidity+ 'unitCode:CAL,SetPoint_Min:-1,SetPoint_Max:1'"
>             }            
>           ]
>         }
>         ]
>     }

I have configured the environment variable 

IOTA_DEFAULT_EXPRESSION_LANGUAGE=jexl

When I send the measurement i am getting as below :
published measurement : {,"Air_Pressure":1143.7,"Humidity":40.64}

 

>   {
>      "id": "urn:ngsi-ld:Motion:test1",
>     "type": "Motion",
>     "Air_Pressure": {
>         "type": "Number",
>         "value": 1143,
>         "metadata": {}
>     },
>     "Humidity": {
>         "type": "Number",
>         "value": 40,
>         "metadata": {}
>     }
> }

Here i have observed that float value converting to integer. Can anyone on this issue.

**Comment by fgalan** [comment]
@chicco785 what do you think?

**Comment by chicco785** [comment]
in theory this is covered by tests:

expression:
https://github.com/telefonicaid/iotagent-node-lib/blob/master/test/unit/ngsiv2/expressions/jexlBasedTransformations-test.js#L118

test:
https://github.com/telefonicaid/iotagent-node-lib/blob/master/test/unit/ngsiv2/expressions/jexlBasedTransformations-test.js#L437

regarding his expression is hard to follow it, i can't understand where the parameters come from :/


Air_Pressure + 'unitCode:CAL,SetPoint_Min:-1,SetPoint_Max:1'

what are  'unitCode:CAL,SetPoint_Min:-1,SetPoint_Max:1' ?
how should be they evaluated ?

did he try without this weird expression which may cause the values to be cast to a string and then back again to a number?

**Comment by achandana** [comment]
Hi chicco I want value as floatnumber+ string  so in above expression i am giving Air_Pressure + 'unitCode:CAL,SetPoint_Min:-1,SetPoint_Max:1' so that i will get  the my desired output.
Is it possible to get the floatnumber and string in value

**Comment by chicco785** [comment]
so, i suspect that Number in your definition is the wrong type, should be "Text". Said so, I am not sure this will solve the issue.

In my understanding assuming that the input is 1143.7, the output you expect is:

1143.7unitCode:CAL,SetPoint_Min:-1,SetPoint_Max:1

**Comment by achandana** [comment]
yes i am expecting the above output  but i am always getting only integer. Decimal values not coming

**Comment by chicco785** [comment]
what is not clear to me if you get the rest of the string or not.

**Comment by achandana** [comment]
Rest of the string is coming but see the below output in Context broker:

    "id": "urn:ngsi-ld:Motion:test1",
    "type": "Motion",
    "Air_Pressure": {
        "type": "Text",
        "value": "944",
        "metadata": {}
    },
    "Humidity": {
        "type": "Text",
        "value": "49unitCode:CAL,SetPoint_Min:-1,SetPoint_Max:1",
        "metadata": {}
    }

**Comment by chicco785** [comment]
ok, so the problem is either:

1. the way the expression is computed:
https://github.com/telefonicaid/iotagent-node-lib/blob/c5cca577a7c132739057a1151fa9dade8f40c355/lib/plugins/jexlParser.js#L125

in this case, the problem is in the library, so we can only open a bug and open they fix it.

2. the way the context passed to the expression is extracted:
https://github.com/telefonicaid/iotagent-node-lib/blob/c5cca577a7c132739057a1151fa9dade8f40c355/lib/plugins/jexlParser.js#L71

I hope is this second case, and probably, a reasonable fix, is just inverting the order of casting from:

```
if (Number.parseInt(attributeList[i].value)) {
          value = Number.parseInt(attributeList[i].value);
        } else if (Number.parseFloat(attributeList[i].value)) {
          value = Number.parseFloat(attributeList[i].value);
        } else if (String(attributeList[i].value) === 'true') {
          value = true;
        } else if (String(attributeList[i].value) === 'false') {
          value = false;
        } else {
          value = attributeList[i].value;
        }
```

to: 

```
if (Number.parseFloat(attributeList[i].value)) {
          value = Number.parseFloat(attributeList[i].value);
        } else if (Number.parseInt(attributeList[i].value)) {
          value = Number.parseInt(attributeList[i].value);
        } else if (String(attributeList[i].value) === 'true') {
          value = true;
        } else if (String(attributeList[i].value) === 'false') {
          value = false;
        } else {
          value = attributeList[i].value;
        }
```

**Comment by achandana** [comment]
Thanks for your great response. I want to give one suggestion to fix this
in the below code first test the type of value and then parse. That would be better like below

> if (isFloat(attributeList[i].value) && Number.parseFloat(attributeList[i].value)) {
>           value = Number.parseFloat(attributeList[i].value);
>         } else if (isInt(attributeList[i].value) && Number.parseInt(attributeList[i].value)) {
>           value = Number.parseInt(attributeList[i].value);
>         } else if (String(attributeList[i].value) === 'true') {
>           value = true;
>         } else if (String(attributeList[i].value) === 'false') {
>           value = false;
>         } else {
>           value = attributeList[i].value;
>         }

**Comment by chicco785** [comment]
i checked the specs of `parseInt` and indeed it works also with float, so that is the issue. this also means that something is weird with the tests that pass.

**Comment by chicco785** [comment]
as regards your suggestion, that would introduce quite an issue with numbers passed as strings, not sure.

**Comment by fgalan** [comment]
Fixed by PR https://github.com/telefonicaid/iotagent-node-lib/pull/869
